ingest/expression_files/dense_ingestor.py

```diff
@@ -0,0 +1,187 @@
+"""Module for ingesting dense matrix files
+
+DESCRIPTION
+Module provides extract and transforms function for gene expression data for
+an dense matrix.
+
+"""
+import collections
+import csv
+import datetime
+import sys
+from typing import List  # noqa: F401
+
+from bson.objectid import ObjectId
+
+try:
+    from expression_files import GeneExpression
+    sys.path.append("../ingest")
+    from ingest_files import IngestFiles
+
+except ImportError:
+    # Used when importing as external package, e.g. imports in single_cell_portal code
+    from .expression_files import GeneExpression
+    sys.path.append("../ingest")
+    from ..ingest_files import IngestFiles
+
+
+class DenseIngestor(GeneExpression, IngestFiles):
+    ALLOWED_FILE_TYPES = ["text/csv",
+                          "text/plain", "text/tab-separated-values"]
+
+    def __init__(self, file_path, study_file_id, study_id, **kwargs):
+        GeneExpression.__init__(self, file_path, study_file_id, study_id)
+        IngestFiles.__init__(
+            self, file_path, allowed_file_types=self.ALLOWED_FILE_TYPES
+        )
+        self.matrix_params = kwargs
+        self.csv_file_handler, self.file_handler = self.open_file(
+            self.file_path)
+        self.gene_names = {}
+        self.header = next(self.csv_file_handler)
+
+    def execute_ingest(self):
+        # Row after header is needed for R format validation
+        row = next(self.csv_file_handler)
+        if DenseIngestor.is_valid_format(self.header, row):
+            # Reset csv reader to first gene row
+            self.csv_file_handler = self.open_file(self.file_path)[0]
+            next(self.csv_file_handler)
+            for gene_docs, data_array_documents in self.transform():
+                self.load(gene_docs, data_array_documents)
+        else:
+            raise ValueError("Dense matrix has invalid format")
```

Comment:
instead of using if/else, I would structure this block as
```
if not DenseIngestor.is_valid_format(self.header, row):
    raise ValueError(...)
self.csv_filehandler = ...
```

Reply:
cf8b889